File: C++/MEDIUM/2058. Find the Minimum and Maximum Number of Nodes Between Critical Points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    vector<int> nodesBetweenCriticalPoints(ListNode *head)
    {
        int firstCritical = -1;
        int lastCritical = -1;
        int lastValue = -1;
        int minPath = 100001;
        int index = 0;

        while (head)
        {
            if (lastValue != -1 && head->next)
            {
                if ((head->val > lastValue && head->val > head->next->val) || (head->val < lastValue && head->val < head->next->val))
                {
                    minPath = lastCritical != -1 ? min(index - lastCritical, minPath) : 100001;
                    firstCritical = firstCritical == -1 ? index : firstCritical;
                    lastCritical = index;
                }
            }
            lastValue = head->val;
            index++;
            head = head->next;
        }

        if (firstCritical == -1 || firstCritical == lastCritical)
        {
            return {-1, -1};
        }
        else
        {
            return {minPath, lastCritical - firstCritical};
        }
    }
};

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        int bef = head.val;
        head = head.next;
        int lo = head.val;
        if (head.next == null)
        {
            return new int[] {-1, -1};
        }
        int nex = head.next.val;

        
        
        int idx = 1;
        
        int lastIdx = -1;
        int firstIdx = -1;
        int[] answer = new int[] {200000, -1};

        while (head.next != null)
        {
            nex = head.next.val;
            if (
                (bef < lo && lo > nex) ||
                (bef > lo && lo < nex)
            )
            {
                System.out.println(idx);
                if (firstIdx == -1)
                {
                    firstIdx = idx;
                }
                if (lastIdx != -1)
                {
                    answer[0] = Math.min(answer[0], idx - lastIdx);
                    answer[1] = Math.max(answer[1], idx - firstIdx);
                }
                lastIdx = idx;
            }
            idx++;
            bef = head.val;
            head=head.next;
            lo = head.val;
        }
        if (answer[0] == 200000)
        {
            answer[0] = -1;
        }

        return answer;
    }
}